Comorphisms/CASL2Skolem.hs

```diff
@@ -0,0 +1,200 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeSynonymInstances, FlexibleInstances #-}
+{- |
+Module      :  $Header$
+Description :  skolemization as an institution comorphism
+Copyright   :  (c) Mihai Codescu, 2016
+License     :  GPLv2 or higher, see LICENSE.txt
+
+Maintainer  :  codescu@iws.cs.uni-magdeburg.de
+Stability   :  provisional
+Portability :  non-portable (imports Logic.Comorphism)
+
+removing existential quantifiers from every formula
+follows http://resources.mpi-inf.mpg.de/departments/rg1/teaching/autrea-ss10/script/lecture10.pdf
+
+-}
+
+module Comorphisms.CASL2Skolem where
+
+import Logic.Logic
+import Logic.Comorphism
+
+import CASL.Logic_CASL
+import CASL.AS_Basic_CASL
+import CASL.Sign
+import CASL.Morphism
+import CASL.Sublogic as SL hiding (bottom)
+import CASL.Induction
+import CASL.Quantification
+
+import Common.Result
+import Common.Id
+import qualified Data.Set as Set
+import Common.AS_Annotation
+import Common.ProofTree
+import qualified Common.Lib.MapSet as MapSet
+
+import qualified Common.Lib.Rel as Rel
+
+import Data.List(partition)
+
+data CASL2Skolem = CASL2Skolem deriving Show
+
+instance Language CASL2Skolem where
+    language_name CASL2Skolem = "CASL2Skolem"
+
+instance Comorphism CASL2Skolem
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree where
+    sourceLogic CASL2Skolem = CASL
+    sourceSublogic CASL2Skolem = SL.caslTop --TODO
+    targetLogic CASL2Skolem = CASL
+    mapSublogic CASL2Skolem sl = Just sl -- TODO
+    map_theory CASL2Skolem = mapTheory
+    map_morphism CASL2Skolem = error "nyi"
```

Comment:
yes, indeed there is a problem. If E = { exists x:s . x=x } and E' = \emptyset, then we have a theory morphism. But the Skolemisation, a=a, cannot be mapped, because there is (say) no constant in the target signature.